File: frontend/src/services/mlService.js
import api from './api';

class MlService {
  /**
   * Upload an Excel file to train the default model (LightGBM / RandomForest).
   * Routes: React → Java /api/ml/train → Python /train
   */
  uploadAndTrain(file) {
    const formData = new FormData();
    formData.append('file', file);
    return api.post('/ml/train', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  }

  /**
   * Train Random Forest AND XGBoost on the same Excel file and compare
   * their metrics side by side.
   *
   * Routes: React → Java /api/ml/compare-models → Python /compare-rf-xgb
   *
   * Response shape:
   *   { random_forest: {r2, rmse, mae}, xgboost: {r2, rmse, mae}, winner, ... }
   */
  compareModels(file) {
    const formData = new FormData();
    formData.append('file', file);
    return api.post('/ml/compare-models', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  }

  /**
   * Predict total demand for the next 12 months.
   * Routes: React → Java /api/ml/predict-year → Python /predict-year
   * @param {string|null} productCode  Optional product code to scope the forecast
   */
  predictYear(productCode = null) {
    const params = productCode ? { product_code: productCode } : {};
    return api.get('/ml/predict-year', { params });
  }
}

export default new MlService();

File: backend/app/src/main/java/com/ibtihel/app/services/ml/MlService.java
package com.ibtihel.app.services.ml;

import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

public interface MlService {

    /**
     * Predict total demand for the next 12 months.
     * @param productCode optional — scope the forecast to a specific product
     */
    Map<String, Object> predictYear(String productCode);

    /**
     * Upload an Excel file and train the default model (LightGBM / RandomForest).
     */
    Map<String, Object> uploadAndTrain(MultipartFile file);

    /**
     * Train Random Forest and XGBoost on the same Excel file, compare their
     * metrics side by side, and save the best model as the active one.
     *
     * Routes: Java → Python POST /compare-rf-xgb
     *
     * @param file Excel file with historical stock data
     * @return comparison result: metrics for both models + winner
     */
    Map<String, Object> compareRfXgb(MultipartFile file);

    /**
     * Send a natural language message to the AI assistant.
     * @param message the user's question in French
     */
    Map<String, Object> chat(String message);
}
